File: js/longest-substring-without-repeating-chars.js
// https://leetcode.com/problems/longest-substring-without-repeating-characters/
// Related Topics: Hash Table, Two Pointers, String, Sliding Window
// Difficulty: Medium

/*
Initial thoughts:
Using a sliding window we start at the beginning of the array
and continuously move the right side of the window to the right,
all the while adding each element to a Set.
If the character is recurring, we start another sliding window
with the left side moved forward by one elemenet.
The size of the biggest possible sliding window is the
longest substring without repeating chars.

Time complexity: O(n^2) where n is the number of elements
Space complexity: O(min(n,m)) where n is the number of elements and
m is the number of characters in the alphabet

*/

/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = s => {
  if (s.length === 0) return 0;
  let maxLength = 0;

  for (let i = 0; i < s.length; i++) {
    const set = new Set([s[i]]);
    for (let j = i + 1; j < s.length; j++) {
      if (set.has(s[j])) break;
      set.add(s[j]);
    }
    maxLength = Math.max(maxLength, set.size);
  }
  return maxLength;
};

/*
Optimization:
Since we are looking for the longest substring
we can stop looping if the left side of the sliding window
has reached a point where the rest of the potential substrings
won't be longer than the already found maxLength.

This won't change the time complexity since in a worst case scenario
we won't find anything, or the longest possible substring until we reach
the end of the loop.

Time complexity: O(n^2) where n is the number of elements
Space complexity: O(min(n, m)) where n is the number of elements and
m is the number of characters in the alphabet
*/

/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = s => {
  if (s.length === 0) return 0;
  let maxLength = 0;

  for (let i = 0; i < s.length - maxLength; i++) {
    const set = new Set([s[i]]);
    for (let j = i + 1; j < s.length; j++) {
      if (set.has(s[j])) break;
      set.add(s[j]);
    }
    maxLength = Math.max(maxLength, set.size);
  }
  return maxLength;
};

/*
Optimization:
In our previous approaches we used to start over checking
every character after the left side of our sliding window
moved forward.
The here is to keep our Set and just remove the left most element
from it when the left side of the sliding window moves forward.
This approach enables us to reuse the already added items to the Set,
decreasing our time complexity by one degree.

Time complexity: O(n) where n is the number of elements in the array
Space complexity: O(min(n,m)) where n is the number of elements in the array
and m is the number of characters in the alphabet
*/

/**
 * @param {string} s
 * @return {number}
 */
const lengthOfLongestSubstring = s => {
  if (s.length === 0) return 0;

  let left = (right = 0);
  let maxLength = 0;
  let set = new Set();
  while (left < s.length && right < s.length) {
    if (!set.has(s[right])) {
      set.add(s[right]);
      maxLength = Math.max(maxLength, set.size);
      right++;
    } else {
      set.delete(s[left]);
      left++;
    }
  }
  return maxLength;
};

/*
Optimization:
Building on the previous approach, we can exit the loop
if the left side of the sliding window becomes greater
than s.length - maxLength because there can't be a potentially
bigger substring.

Time complexity: O(n) where n is the number of elements in the array
Space complexity: O(min(n,m)) where n is the number of elements in the array
and m is the number of characters in the alphabet
*/

/**
 * @param {string} s
 * @return {number}
 */

const lengthOfLongestSubstring = s => {
  if (s.length === 0) return 0;

  let left = (right = 0);
  let maxLength = 0;
  const set = new Set();
  while (left < s.length - maxLength && right < s.length) {
    if (!set.has(s[right])) {
      set.add(s[right]);
      maxLength = Math.max(maxLength, set.size);
      right++;
    } else {
      set.delete(s[left]);
      left++;
    }
  }
  return maxLength;
};

/*
Optimization:
Instead of using a Set we could use a Map and keeping track of
each characters index.
Now, when we encounter a duplicated character, instead of moving
the left side of the window by one, we could jump to the position
where the repeating character is + one.
This way, we are going to decrease the time complexity from O(2n) to O(n) which
isn't actually an optimization in terms of Big O, but it's still a small gain.

Time complexity: O(n) where n is the number of elements in the array
Space complexity: O(min(n,m)) where n is the number of elements in the array
and m is the number of characters in the alphabet
*/

/**
 * @param {string} s
 * @returns {number}
 */

const lengthOfLongestSubstring = s => {
  if (s.length === 0) return 0;

  let maxLength = 0;
  const map = new Map();

  for (let j = 0, i = 0; j < s.length; j++) {
    if (map.has(s[j])) {
      i = Math.max(map.get(s[j]), i);
    }
    maxLength = Math.max(maxLength, j - i + 1);
    map.set(s[j], j + 1);
  }
  return maxLength;
};

File: python/longest-substring-without-repeating-chars.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
# Related Topics: Hash Table, Two Pointers, String, Sliding Window
# Difficulty: Medium


class Solution:
    # Time complexity: O(n) where n is the length of the input string
    # Space complexity: O(1) since the dictionary has never more than 26 elements
    # Note: This is an optimized version of the further down algorithm
    # Since it does only one pass
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_sub = 0
        left = 0
        have = dict()
        
        for right, el in enumerate(s):
            if el in have:
                left = max(left, have[el]+1)
            max_sub = max(max_sub, right - left + 1)
            have[el] = right
        return max_sub
    
    
    # Time complexity: O(n) where n is the length of the input string
    # Space complexity: O(1) since the set has never more than 26 elements
    def lengthOfLongestSubstring2(self, s: str) -> int:
        max_sub = 0
        left = 0
        have = set()
        
        for right, el in enumerate(s):
            while el in have:
                have.remove(s[left])
                left += 1
            have.add(el)
            max_sub = max(max_sub, right - left + 1)
        return max_sub

# Initial Thoughts:
# Using a Set and two pointers we are going to create a sliding window
# that stores all its values in the Set.
# Initially we start by advancing the right pointer and add the chars to
# the set. Each time we calculate the distance between the two pointers as
# the max substring without recurring chars and store it.
# Whenever we encounter a duplicate value, we remove the char that the left
# pointer is pointing to from the Set and move it forward until there is
# no duplicate. We repeat this until the left's distance to the end is no more
# than the max substring that we already found.

# Time complexity O(n) where n == len(s)
# Space complexity O(min(n,m)) where n == len(s) and m == len(charactersInAlphabet)
# We could also argue that the space complexity is constant since the number of characters
# in the alphabet is constant and well defined


class Solution2:
    def lengthOfLongestSubstring(self, s: str) -> int:
        mySet = set()
        max = 0
        left = right = 0
        while left < len(s)-max:
            if not s[right] in mySet:
                mySet.add(s[right])
                right += 1
                max = right - left if right - left > max else max
            else:
                mySet.remove(s[left])
                left += 1

        return max
